File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/SurfaceViewEngine.java
package kaufland.com.swipelibrary.dragengine;

import android.graphics.Rect;
import android.view.View;

import kaufland.com.swipelibrary.SurfaceView;
import kaufland.com.swipelibrary.SwipeDirectionDetector;
import kaufland.com.swipelibrary.SwipeLayout;
import kaufland.com.swipelibrary.SwipeResult;
import kaufland.com.swipelibrary.SwipeState;
import kaufland.com.swipelibrary.SwipeViewLayouter;

import static kaufland.com.swipelibrary.SwipeDirectionDetector.SWIPE_DIRECTION_LEFT;
import static kaufland.com.swipelibrary.SwipeDirectionDetector.SWIPE_DIRECTION_RIGHT;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.CLOSED;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.LEFT_OPEN;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.RIGHT_OPEN;

/**
 * Created by sbra0902 on 29.03.17.
 */

public class SurfaceViewEngine implements DraggingEngine<SurfaceView> {

    public static final double AUTO_OPEN_SPEED_TRESHOLD = 800.0;
    public static final double LEFT_FULL_AUTO_OPEN_TRESHOLD = AUTO_OPEN_SPEED_TRESHOLD * 2;
    public static final double RIGHT_FULL_AUTO_OPEN_TRESHOLD = -LEFT_FULL_AUTO_OPEN_TRESHOLD;

    private SwipeViewLayouter mLayouter;

    private SurfaceView mSurfaceView;

    private DraggingEngine mLeftDragView;

    private DraggingEngine mRightDragView;

    public SurfaceViewEngine(SwipeViewLayouter layouter) {
        mLayouter = layouter;
    }

    @Override
    public void moveView(float offset, SurfaceView view, View changedView) {

        if(changedView.equals(mSurfaceView)){
            return;
        }

        if(changedView.equals(mLeftDragView.getDragView())){
            mSurfaceView.setX(mLeftDragView.getDragView().getX() + mLeftDragView.getWidth());
        }

        if(changedView.equals(mRightDragView.getDragView())){
            mSurfaceView.setX(mRightDragView.getDragView().getX() - mSurfaceView.getWidth());
        }
    }

    @Override
    public SwipeResult determineSwipeHorizontalState(float velocity, SwipeDirectionDetector swipeDirectionDetector, SwipeState swipeState, final SwipeLayout.SwipeListener mSwipeListener, View releasedChild) {

        if(!mSurfaceView.equals(releasedChild)){
            return null;
        }

        int swipeDirection = swipeDirectionDetector.getSwipeDirection();

        if (swipeState.getState() == SwipeState.DragViewState.CLOSED) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {
                if (!mLeftDragView.isDraggable()) {
                    return new SwipeResult(0);
                }

                swipeState.setState(SwipeState.DragViewState.LEFT_OPEN);
                if (velocity > LEFT_FULL_AUTO_OPEN_TRESHOLD || Math.abs(swipeDirectionDetector.getDifX()) > (mLeftDragView.getDragDistance() / 2)) {
                    return new SwipeResult(mLeftDragView.getDragDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(LEFT_OPEN, true);
                        }
                    });
                } else {
                    return new SwipeResult(mLeftDragView.getIntermmediateDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(LEFT_OPEN, mLeftDragView.getIntermmediateDistance() == mLeftDragView.getDragDistance());
                        }
                    });
                }

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {
                if (!mRightDragView.isDraggable()) {
                    return new SwipeResult(0);
                }

                swipeState.setState(RIGHT_OPEN);

                if (velocity < RIGHT_FULL_AUTO_OPEN_TRESHOLD || Math.abs(swipeDirectionDetector.getDifX()) > (mRightDragView.getDragDistance() / 2)) {
                    return new SwipeResult(-mRightDragView.getDragDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(RIGHT_OPEN, true);
                        }
                    });
                } else {
                    return new SwipeResult(-mRightDragView.getIntermmediateDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(RIGHT_OPEN, mRightDragView.getIntermmediateDistance() == mRightDragView.getDragDistance());
                        }
                    });
                }
            }
        } else if (swipeState.getState() == SwipeState.DragViewState.LEFT_OPEN) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {
                swipeState.setState(LEFT_OPEN);

                return new SwipeResult(mLeftDragView.getIntermmediateDistance(), new Runnable() {
                    @Override
                    public void run() {
                        mSwipeListener.onBounce(LEFT_OPEN);
                    }
                });

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {
                swipeState.setState(CLOSED);
                return new SwipeResult(0, new Runnable() {
                    @Override
                    public void run() {
                        mSwipeListener.onSwipeClosed(CLOSED);
                    }
                });
            }
        } else if (swipeState.getState() == RIGHT_OPEN) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {

                swipeState.setState(SwipeState.DragViewState.CLOSED);
                return new SwipeResult(0, new Runnable() {
                    @Override
                    public void run() {
                        mSwipeListener.onSwipeClosed(CLOSED);
                    }
                });

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {
                swipeState.setState(RIGHT_OPEN);
                if (velocity < RIGHT_FULL_AUTO_OPEN_TRESHOLD || Math.abs(swipeDirectionDetector.getDifX()) > (mRightDragView.getDragDistance() / 2)) {
                    return new SwipeResult(-mRightDragView.getDragDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(RIGHT_OPEN, true);
                        }
                    });
                } else {
                    return new SwipeResult(-mRightDragView.getIntermmediateDistance(), new Runnable() {
                        @Override
                        public void run() {
                            mSwipeListener.onSwipeOpened(RIGHT_OPEN, mRightDragView.getIntermmediateDistance() == mRightDragView.getDragDistance());
                        }
                    });
                }
            }

        }

        return new SwipeResult(0);
    }

    @Override
    public void initializePosition(SwipeViewLayouter.DragDirection orientation) {

        mSurfaceView = mLayouter.getSurfaceView();

        mLeftDragView = mLayouter.getDragViewEngineByPosition(SwipeLayout.LEFT_DRAG_VIEW);
        mRightDragView = mLayouter.getDragViewEngineByPosition(SwipeLayout.RIGHT_DRAG_VIEW);
    }

    @Override
    public void moveToInitial() {
        mLayouter.getSurfaceView().setX(0);
    }

    @Override
    public int clampViewPositionHorizontal(View child, int left) {
        boolean isOutsideRightRangeAndBounceNotPossible = left < - mRightDragView.getDragDistance() && !mRightDragView.isBouncePossible();
        boolean isOutsideLeftRangeAndBounceNotPossible = left > mRightDragView.getDragDistance() && !mRightDragView.isBouncePossible();


        if (isOutsideLeftRangeAndBounceNotPossible) {
            return mRightDragView.getDragDistance();
        }

        if(isOutsideRightRangeAndBounceNotPossible){
            return mRightDragView.getDragDistance();
        }


        return left;
    }


    @Override
    public void restoreState(SwipeState.DragViewState state, SurfaceView view) {
        //TODO
//        switch (state) {
//            case LEFT_OPEN:
//                moveToInitial();
//                break;
//            case RIGHT_OPEN:
//                moveView(mDragDistance, view, null);
//                break;
//            case TOP_OPEN:
//                //TODO Implementation
//                break;
//            case BOTTOM_OPEN:
//                //TODO Implementation
//                break;
//            default:
//
//                moveToInitial();
//                break;
//        }
    }

    @Override
    public int getWidth() {
        return mSurfaceView.getWidth();
    }

    @Override
    public int getDragDistance() {
        return 0;
    }

    @Override
    public void forceLayout() {
        mSurfaceView.forceLayout();
    }

    @Override
    public boolean isBouncePossible() {
        return false;
    }

    @Override
    public int getId() {
        return mSurfaceView.getId();
    }

    @Override
    public boolean isDraggable() {
        return false;
    }

    @Override
    public int getIntermmediateDistance() {
        return 0;
    }


    @Override
    public SurfaceView getDragView() {
        return mSurfaceView;
    }
}

File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/SurfaceViewEngine.kt
package kaufland.com.swipelibrary.dragengine

import android.view.View

import kaufland.com.swipelibrary.DragView
import kaufland.com.swipelibrary.SurfaceView
import kaufland.com.swipelibrary.SwipeDirectionDetector
import kaufland.com.swipelibrary.SwipeDirectionDetector.Companion.SWIPE_DIRECTION_LEFT
import kaufland.com.swipelibrary.SwipeDirectionDetector.Companion.SWIPE_DIRECTION_RIGHT
import kaufland.com.swipelibrary.SwipeLayout
import kaufland.com.swipelibrary.SwipeResult
import kaufland.com.swipelibrary.SwipeState
import kaufland.com.swipelibrary.SwipeState.DragViewState.*
import kaufland.com.swipelibrary.SwipeViewLayouter

import kotlin.math.abs

/**
 * Created by sbra0902 on 29.03.17.
 */

class SurfaceViewEngine(private val mLayouter: SwipeViewLayouter) : DraggingEngine<SurfaceView> {

    override var dragView: SurfaceView? = null

    private var mLeftDragViewEngine: DraggingEngine<*>? = null

    private var mRightDragViewEngine: DraggingEngine<*>? = null

    override val dragDistance: Int
        get() = 0

    override val intermediateDistance: Int
        get() = 0

    override val openOffset: Int
        get() = 0

    override fun moveView(offset: Float, view: SurfaceView, changedView: View) {

        if (changedView == dragView) {
            return
        }

        if (mLeftDragViewEngine != null && changedView == mLeftDragViewEngine!!.dragView) {
            dragView?.x = mLeftDragViewEngine!!.dragView!!.x + mLeftDragViewEngine!!.dragView!!.width
        }

        if (mRightDragViewEngine != null && changedView == mRightDragViewEngine!!.dragView) {
            dragView?.x = mRightDragViewEngine!!.dragView!!.x - dragView!!.width
        }
    }

    private fun isMinimumDifReached(distanceToNextState: Int, detector: SwipeDirectionDetector): Boolean {
        return abs(distanceToNextState / 2) <= abs(detector.difX) && abs(detector.difX) > abs(detector.difY) && !detector.isHorizontalScrollChangedWhileDragging
    }

    override fun determineSwipeHorizontalState(velocity: Float, swipeDirectionDetector: SwipeDirectionDetector, swipeState: SwipeState, swipeListener: SwipeLayout.SwipeListener, releasedChild: View): SwipeResult? {

        if (dragView != releasedChild) {
            return null
        }

        val swipeDirection = swipeDirectionDetector.swipeDirection

        if (swipeState.state == CLOSED) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {
                if (mLeftDragViewEngine == null || !getDragViewForEngine(mLeftDragViewEngine)!!.isDraggable || !isMinimumDifReached(mLeftDragViewEngine!!.intermediateDistance, swipeDirectionDetector)) {
                    return SwipeResult(0)
                }

                swipeState.state = LEFT_OPEN
                return if (velocity > LEFT_FULL_AUTO_OPEN_THRESHOLD || abs(swipeDirectionDetector.difX) > mLeftDragViewEngine!!.dragDistance / 2) {
                    SwipeResult(mLeftDragViewEngine!!.dragDistance, Runnable { swipeListener.onSwipeOpened(LEFT_OPEN, true) })
                } else {
                    SwipeResult(mLeftDragViewEngine!!.intermediateDistance, Runnable { swipeListener.onSwipeOpened(LEFT_OPEN, mLeftDragViewEngine!!.intermediateDistance == mLeftDragViewEngine!!.dragDistance) })
                }

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {
                if (mRightDragViewEngine == null || !getDragViewForEngine(mRightDragViewEngine)!!.isDraggable || !isMinimumDifReached(mRightDragViewEngine!!.intermediateDistance, swipeDirectionDetector)) {
                    return SwipeResult(0)
                }

                swipeState.state = RIGHT_OPEN

                return if (velocity < RIGHT_FULL_AUTO_OPEN_THRESHOLD || abs(swipeDirectionDetector.difX) > mRightDragViewEngine!!.dragDistance / 2) {
                    SwipeResult(-mRightDragViewEngine!!.dragDistance, Runnable { swipeListener.onSwipeOpened(RIGHT_OPEN, true) })
                } else {
                    SwipeResult(-mRightDragViewEngine!!.intermediateDistance, Runnable { swipeListener.onSwipeOpened(RIGHT_OPEN, mRightDragViewEngine!!.intermediateDistance == mRightDragViewEngine!!.dragDistance) })
                }
            }
        } else if (swipeState.state == LEFT_OPEN) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {
                swipeState.state = LEFT_OPEN

                return SwipeResult(mLeftDragViewEngine!!.intermediateDistance, Runnable { swipeListener.onBounce(LEFT_OPEN) })

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {

                swipeState.state = CLOSED
                return SwipeResult(0, Runnable { swipeListener.onSwipeClosed(CLOSED) })
            }
        } else if (swipeState.state == RIGHT_OPEN) {
            if (swipeDirection == SWIPE_DIRECTION_RIGHT) {

                swipeState.state = CLOSED
                return SwipeResult(0, Runnable { swipeListener.onSwipeClosed(CLOSED) })

            } else if (swipeDirection == SWIPE_DIRECTION_LEFT) {
                swipeState.state = RIGHT_OPEN
                return if (velocity < RIGHT_FULL_AUTO_OPEN_THRESHOLD || abs(swipeDirectionDetector.difX) > mRightDragViewEngine!!.dragDistance / 2) {
                    SwipeResult(-mRightDragViewEngine!!.dragDistance, Runnable { swipeListener.onSwipeOpened(RIGHT_OPEN, true) })
                } else {
                    SwipeResult(-mRightDragViewEngine!!.intermediateDistance, Runnable { swipeListener.onSwipeOpened(RIGHT_OPEN, mRightDragViewEngine!!.intermediateDistance == mRightDragViewEngine!!.dragDistance) })
                }
            }

        }

        return SwipeResult(0)
    }

    override fun initializePosition(orientation: SwipeViewLayouter.DragDirection) {

        dragView = mLayouter.surfaceView

        mLeftDragViewEngine = mLayouter.getDragViewEngineByPosition(SwipeLayout.LEFT_DRAG_VIEW)
        mRightDragViewEngine = mLayouter.getDragViewEngineByPosition(SwipeLayout.RIGHT_DRAG_VIEW)
    }

    override fun clampViewPositionHorizontal(child: View, left: Int): Int {

        if (mRightDragViewEngine == null) {
            if (left <= 0) {
                return 0
            }
        } else {
            val isOutsideRightRangeAndBounceNotPossible = left < -mRightDragViewEngine!!.dragDistance && !getDragViewForEngine(mRightDragViewEngine)!!.isBouncePossible
            if (isOutsideRightRangeAndBounceNotPossible) {
                return -mRightDragViewEngine!!.dragDistance
            }
        }


        if (mLeftDragViewEngine == null) {
            if (left <= -mRightDragViewEngine!!.dragDistance) {
                return -mRightDragViewEngine!!.dragDistance
            }
        } else {
            val isOutsideLeftRangeAndBounceNotPossible = left > mLeftDragViewEngine!!.dragDistance && !getDragViewForEngine(mLeftDragViewEngine)!!.isBouncePossible


            if (isOutsideLeftRangeAndBounceNotPossible) {
                return mLeftDragViewEngine!!.dragDistance
            }
        }


        return left
    }


    override fun restoreState(state: SwipeState.DragViewState, view: SurfaceView) {

        when (state) {
            LEFT_OPEN -> dragView?.x = mLeftDragViewEngine!!.dragDistance.toFloat()
            RIGHT_OPEN -> dragView?.x = (-mRightDragViewEngine!!.dragDistance).toFloat()
            TOP_OPEN -> {
            }
            BOTTOM_OPEN -> {
            }
            else -> dragView?.x = 0f
        }//TODO Implementation
        //TODO Implementation
    }

    private fun getDragViewForEngine(engine: DraggingEngine<*>?): DragView? {
        return if (engine != null) engine.dragView as DragView else null
    }

    companion object {

        private const val AUTO_OPEN_SPEED_THRESHOLD = 800.0
        private const val LEFT_FULL_AUTO_OPEN_THRESHOLD = AUTO_OPEN_SPEED_THRESHOLD * 2
        private const val RIGHT_FULL_AUTO_OPEN_THRESHOLD = -LEFT_FULL_AUTO_OPEN_THRESHOLD
    }
}
